### PR TITLE
Fix handleLoadError for async cache errors

### LINKED ISSUE #460: Returning false in CacheErrorHandler's handleLoadError() does not invoke original method
### Background

Using a cache server in a rapidly developing system can lead to SerializationExceptions if old cached entries are no longer compatible with newly introduced changes to the corresponding models.
To handle this issue, I replaced the `DefaultCacheErrorHandler` and `AsyncCacheErrorHandler` with my own custom implementation. It's only purpose is to handle SerializationExceptions in the `handleLoadError` method by invalidating the cached entry and returning `false`, which should lead to the invocation of the original method.

### Behavior

Returning `false` in the CacheErrorHandler's `handleLoadError` method should invoke the original method.

### Actual Behaviour

Even though I can see that the cached entry is invalidated, the original method is not invoked, at least not for cached methods which return a reactive type (`Mono`in my case).

I think I also found the cause of this issue. The `asyncCacheGet` function in the CacheInterceptor defines a default of `null` in case the `asyncCache.get()` finishes exceptionally. This means that the surrounding CompletableFuture will eventually return null. The problem arises when the CompletableFuture is transformed into a Mono in the corresponding `interceptSingle` method.
Passing null into a Mono will always lead to a `Mono.empty()` situation which will in turn abort the entire reactive chain. Because of this, the defined `flatMap` which triggers the invocation of the original method is never executed.

### Steps To Reproduce

The above might be a very complicated way of explaining things, so let me try again with this minimal example.
A very condensed code of what the `interceptSingle` and `asyncCacheGet` actually do could look like the following (kotlin code):
```kotlin
fun testCacheErrorFallback(fallback: Optional<String>?): Mono<String> {
  val future = CompletableFuture
    .failedFuture<Optional<String>>(RuntimeException("Expected exception"))
    .exceptionally { fallback }

  return Mono
    .defer { Mono.fromCompletionStage(future) }
    .flatMap { result ->
      if (result.isPresent) {
        // simulates cached entry found
        Mono.just(result.get())
      } else {
        // underlying method has to be invoked
        Mono.just("optional result was missing")
      }
    }
    .switchIfEmpty(Mono.just("mono was empty"))
}
```
The `fallback` parameter in the above example represents the default the actual `asyncCacheGet` method returns in case of an error.

Now if we call this kotlin function, one of two things can happen:
```kotlin
testCacheErrorFallback(Optional.empty())
  .doOnNext { log.info(it) }
  .subscribe() // logs "optinal result was missing"
testCacheErrorFallback(null)
  .doOnNext { log.info(it) }
  .subscribe() // logs "mono was empty"
```

### Solution
In my eyes there are two possible solutions.
- Either the `asyncCacheGet` method returns Optional.empty() in case of an error, or
- `interceptsingle` needs to add an additional `switchIfEmpty` reactive step

I would assume that the first option is not only easier but actually correct because the `asyncCacheGet` already defines its return type as `CompletableFuture<? extends Optional<?>>`.

Did I miss something?
If not then I'm happy to open a PR by implementing my first suggested solution.

### Environment Information

- JDK 11
- Windows and Linux

### Example Application

_No response_

### Version

3.3.0 (micronaut-cache)

### PR DESCRIPTION
**Pull request by timyates**
Closes #460

The fix is to change asyncCacheGet to use Optional.empty() as a default in case of exception, instead of null.

As part of this PR, I have also switched to use CompletableFuture.exceptionallyAsync which is from Java 12 instead of our own

### CODE REVIEW

File: cache-core/src/main/java/io/micronaut/cache/interceptor/CacheInterceptor.java [hunk 1]
```diff
@@ -852,13 +857,11 @@ private CompletableFuture<? extends Optional<?>> asyncCacheGet(AsyncCache<?> asy
...
-                                       exceptionallyAsync(throwable,
-                                                          () -> errorHandler.handleLoadError(asyncCache,
-                                                                                             key,
-                                                                                             asRuntimeException(throwable)),
-                                                          null));
+        return asyncCache
+            .get(key, requiredType)
+            .exceptionallyAsync(throwable ->
+                exceptionallyAsync(throwable, errorHandler.handleLoadError(asyncCache, key, asRuntimeException(throwable)), Optional.empty())
+            );
```
Comment (timyates) [inline cache-core/src/main/java/io/micronaut/cache/interceptor/CacheInterceptor.java:864]: This is the change, we go from having `null` as a default return value to `Optional.empty`.

**Review by sdelamo — APPROVED** [review]
